text/0000-proc-macro-expansion-api.md

```diff
@@ -0,0 +1,379 @@
+
+```rust
+struct S; // [Def]
+struct T; // [Call]
+
+// [Call]
+//   v
+impl T {
+    // These implementation functions can refer to S because
+    // they're in the same scope
+    ... // [Def]
+}
+```
+
+* If the `[Call]` scope of `helper!` is the `[Def]` scope of `my_proc!`, then `helper!` will 'export' or 'expose' the declaration of `T` to `my_proc!`, which lets `my_proc!` refer to `T`. This lets us delegate part of the implementation of `my_proc!` to other proc and decl macros (perhaps from other crates).
+
+* If instead the `[Call]` scope of `helper!` is the `[Call]` scope of `my_proc!`, then `helper!` will export the declarations to the caller of `my_proc!` instead of `my_proc!`. If we don't need access to `T` and just want to export it straight to the caller of `my_proc!` (or if `helper!` is actually just part of the caller's input to `my_proc!`, like `my_proc!(helper!(...))`) then this is what we want.
+
+Since both of these are legitimate use cases, `MacroCall` provides `call_from` to set what the `[Call]` scope of the macro call will be.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The proposed additions to the proc macro API in `proc_macro` are outlined above in the [API overview](#api-overview). Here we focus on technical challenges.
+
+When a source file is parsed any `macro_rules!` and `macro` definitions get added to a definition map long before the first macro is expanded. Procedural macros currently need to live in a separate crate, and it seems they will for a while. This means that _in principle_ any macro call that would resolve in the caller's scope should be available to resolve at the time the proc macro is expanded.
+
+Built-in macros already look more and more like proc macros (or at the very least could be massaged into acting like them), and so they can also be added to the definition map.
+
+Since proc macros and `macro` definitions are relative-path-addressable, the proc macro call context needs to keep track of what the path was at the call site. I'm not sure if this information is available at expansion time, but are there any issues getting it?
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal:
+
+* Increases the API surface of `proc_macro` and any crate trying to emulate it. In fact, since it requires actually evaluating macro calls it isn't clear how a third-party crate like `proc_macro2` could even try to emulate it.
+
+* Greatly increases the potential for hairy interactions between macro calls. This opens up more of the implementation to be buggy (that is, by restricting how macros can be expanded, we might keep implementation complexity in check).
+
+* Relies on proc macros being in a separate crate, as discussed in the reference level explanation [above](#reference-level-explanation). This makes it harder to implement any future plans of letting proc macros be defined and used in the same crate.
```

Comment:
I'd like to highlight this drawback. Are the gains in this RFC enough to outweigh this drawback?

Reply:
Indeed, why does it require a separate crate for proc macros? Can you elaborate?

Reply:
Thinking about it more, this expansion API doesn't add any _extra_ constraints to where a proc macro can be defined, so I guess this shouldn't really be here.

Originally I was worried about macro name resolution (I thought having proc macros in a separate crate at the call site would make that easier but given that there are other issues involving macro paths this seems redundant to worry about), and collecting definitions in an 'executable' form.

Declarative macros can basically be run immediately after they're parsed because they're all compositions of pre-existing built-in purely-syntactic compiler magic. Same-crate procedural macros would need to be 'pre-compiled' like they're tiny little inline `build.rs`s scattered throughout your code. I thought this would interact poorly in situations line this:

```rust
#[macro_use]
extern crate some_crate;

#[proc_macro]
fn my_proc_macro(ts: TokenStream) -> TokenStream { ... }

fn main() {
    some_crate::a_macro!(my_proc_macro!(foo));
}
```

How does `some_crate::a_macro!` know how to expand `my_proc_macro!`?

In hindsight, this is just a roundabout way of hitting an existing problem with same-crate proc macros:

```rust
// Not a proc-macro.
fn helper(ts: TokenStream) -> TokenStream { ... }

#[proc_macro]
fn a_macro(ts: TokenStream) -> TokenStream {
    let helped_ts = helper(ts);
    ...
}

fn main() {
    a_macro!(foo);
}
```

Same question: how does `a_macro!` know how to evaluate `helper`? I think whatever answer we find there will translate to this macro expansion problem.

Anyway, I'm now slightly more confident that that particular drawback isn't _introduced_ by this RFC. Should I remove it?

Reply:
Yeah, I'd tend to agree with that assessment. Is there an RFC open for same-crate proc macros currently? If so, I'd be curious to read it over.

Reply:
I remember reading some fleeting comments about it, but I just had a quick look around and I can't find anything about plans for it.

Reply:
I'm no expert wrt. proc macros.. I'd also be interested in any resources wrt. same-crate macros.

Thanks for the detailed review and changes =)

Reply:
@pierzchalski On a related note, my WIP PR can be found here: <https://github.com/rust-lang/rust/pull/47992#issuecomment-363155321>. I'm going to make another big commit & push in an hour I think.